### ISSUE TITLE
slices from collate_fn returns unexpected result

### ISSUE THREAD

**Issue by minhtriet** [pyg-team/pytorch_geometric#1901]
I am trying to create a custom dataset. My `dataset.data` is `Data(edge_index=[2, 93003], edge_to_id=[1], test_mask=[93003], train_mask=[93003], val_mask=[93003], vertice_to_id=[1], y=[93003])`. The number in square brace is `shape`

However, `dataset.slice` is `{'edge_index': tensor([0, 93003]), 'y': tensor([0, 93003]), 'train_mask': tensor([0, 93003]), 'val_mask': tensor([0, 93003]), 'test_mask': tensor([0, 93003])}`. The number in `([...])` is actual value. Therefore, I cannot get an item with index larger than 2, while I have 93003 datapoint.

I have seen that slices typically has a longer length. What did I do wrong? My code of creating the dataset is

```python
edge_index = torch.LongTensor(edge_index)
data = Data(edge_index=edge_index.t().contiguous())
data.y = torch.tensor(edge_type)
data.train_mask = torch.cat((torch.ones(n_train, dtype=torch.bool), torch.zeros(n_val+n_test, dtype=torch.bool)))
data.val_mask = torch.cat((torch.zeros(n_train, dtype=torch.bool), torch.ones(n_val, dtype=torch.bool), torch.zeros(n_test, dtype=torch.bool)))
data.test_mask = torch.cat((torch.zeros(n_train+n_val, dtype=torch.bool), torch.ones(n_test, dtype=torch.bool)))
collated_data, slices = self.collate([data])
torch.save((collated_data, slices), *self.processed_paths)
```

**Comment by rusty1s** [comment]
You are collating a single `data` object, so your dataset will only contain a single object. If you want to provide more data objects by the dataset, you need to do:
```python
self.collate([data_1, data_2, ...., data_N])
```

**Comment by minhtriet** [comment]
Hi thanks for the answer. I change my code to
```python
data_list = [Data(edge_index=edge_index[:, _], y=edge_type[_]) 
            for _ in range(edge_index.shape[1])]
collated_data, slices = self.collate(data_list)
```
The `slices` seems to work, but in the following line in `in_memory_dataset.py` changes my `data[edge_index]` from a list of 93303 tensors `2*1` tensor to a tensor with shape 1860006.
```python 
    for key in keys:
            item = data_list[0][key]
            if torch.is_tensor(item):
                data[key] = torch.cat(data[key],
                                      dim=data.__cat_dim__(key, item))
            elif isinstance(item, int) or isinstance(item, float):
                data[key] = torch.tensor(data[key])
```
Seems like I misunderstood you. What should I write instead?

**Comment by rusty1s** [comment]
I'm not sure what you are trying to achieve. Do you have multiple graphs in your dataset or does it simply contain a single graph?
In case of the latter, your initial code was already correct. In case you want to include multiple graphs in your dataset, you create a separate `data` object for each of those graphs, and use `self.collate` to convert it into the `InMemoryDataset` format:
```python
def process(self):
    data_list = [] 
    for i in range(num_graphs):
        data = Data(...)  # Create graph i
        data_list.append(data)
    data, slices = self.collate(data_list)
    torch.save((data, slices), path)
```

**Comment by minhtriet** [comment]
> I'm not sure what you are trying to achieve. Do you have multiple graphs in your dataset or does it simply contain a single graph?
That is correct, I have one graph, some links are for training, some for validation and testing

The initial code runs without a problem. However, when I tried 
```
>>> dataset = WordNet18RR('.')
>>> dataset.__getitem__(0)
Data(edge_index=[2, 93003], test_mask=[93003], train_mask=[93003], val_mask=[93003], y=[93003])
>>> dataset.__getitem__(1)
{IndexError}index 2 is out of bounds for dimension 0 with size 2
```
The same happens if I use `DataLoader`. Somehow the only datapoint contains the _shape_ of the dataset, not the actual data. When I debugged it, seems like because `slices[key].append(s)` in `in_memory_dataset.py` only append the size of the dataset. `dataset.data` gives correct answer.

**Comment by rusty1s** [comment]
Yeah, `WordNet` does contain only a single graph, so accessing `dataset[1]` yields an index error. I'm not sure what you want `slices` to look like, but internally it is just use to split datasets into single data objects.

**Comment by minhtriet** [comment]
Ah, I see. Thank you.

Just wanted to sample edges from `edge_index` with `DataLoader` to do `edge_type` prediction. I planned to use `DataLoader(dataset)`, but seems like it is wrong, perhaps something like `DataLoader(dataset.edge_index)`.

If you don't mind, I would also create a merge request for WN18RR dataset.

**Comment by rusty1s** [comment]
We provide numerous ways to sample information from a single graph, e.g., via `NeighborSampler`, `Cluster-GCN` or `GraphSAINT`. However, your graph is not that big, so I would recommend to just train on it in a full-batch fashion.

I always appreciate PRs, although I'm not sure how `WordNet18RR` differs from the one we already provide: `torch_geometric.datasets.WordNet18`.

**Comment by minhtriet** [comment]
> In addition, it has been noted that the WN18 andFB15k datasets suffer from test set leakage, due to inverse relations from the training set being present in the test set – however, the extent of this issue has so far not been quantified.

... but enough for them to create a new dataset. [Source](https://arxiv.org/abs/1707.01476)

**Comment by minhtriet** [comment]
[It is done](https://github.com/rusty1s/pytorch_geometric/pull/1910)
